wremnants/CardTool.py

```diff
@@ -715,7 +731,774 @@ def writeOutput(self, args=None, xnorm=False, forceNonzero=True, check_systs=Tru
+            logk_sparse_dense_shape = (norm_sparse_indices.shape[0], 2*nsyst)
+            logk_sort_indices = np.argsort(np.ravel_multi_index(np.transpose(logk_sparse_indices),logk_sparse_dense_shape))
+            logk_sparse_indices = logk_sparse_indices[logk_sort_indices]
+            logk_sparse_values = logk_sparse_values[logk_sort_indices]
+            logk_sort_indices = None
+
+        else:
+            logger.info(f"Write out dense array")
+            #initialize with zeros, i.e. no variation
+            norm = np.zeros([nbinsfull,nproc], dtype)
+            logk = np.zeros([nbinsfull,nproc,2,nsyst], dtype)
+
+            ibin = 0
+            for nbinschan, chan in zip(ibins, chans):
+                dict_norm_chan = dict_norm[chan]
+                dict_logkavg_chan = dict_logkavg[chan]
+                dict_logkhalfdiff_chan = dict_logkhalfdiff[chan]
+
+                for iproc, proc in enumerate(procs):
+                    if proc not in dict_norm_chan:
+                        continue
+                    norm[ibin:ibin+nbinschan, iproc] = dict_norm_chan[proc]
+
+                    dict_logkavg_proc = dict_logkavg_chan[proc]
+                    dict_logkhalfdiff_proc = dict_logkhalfdiff_chan[proc]
+                    for isyst, syst in enumerate(systs):
+                        if syst not in dict_logkavg_proc.keys():
+                            continue
+
+                        logk[ibin:ibin+nbinschan,iproc,0,isyst] = dict_logkavg_proc[syst]
+                        logk[ibin:ibin+nbinschan,iproc,1,isyst] = dict_logkhalfdiff_proc[syst]
+                        
+                ibin += nbinschan
+
+
+        #free memory
+        # logkavg_proc = None
+        # logkhalfdiff_proc = None
+            
+        #compute poisson parameter for Barlow-Beeston bin-by-bin statistical uncertainties
+        kstat = np.square(sumw)/sumw2
+        #numerical protection to avoid poorly defined constraint
+        kstat = np.where(np.equal(sumw,0.), 1., kstat)
+
+        #write results to hdf5 file
+        procSize = nproc*np.dtype(dtype).itemsize
+        systSize = 2*nsyst*np.dtype(dtype).itemsize
+        amax = np.amax([procSize,systSize])
+        if amax > chunkSize:
+            logger.warning(f"Maximum chunk size in bytes was increased from {chunkSize} to {amax} to align with tensor sizes and allow more efficient reading/writing.")
+            chunkSize = amax
+
+        #create HDF5 file (chunk cache set to the chunk size since we can guarantee fully aligned writes
+        outfilename = self.cardName.replace('_{chan}','').replace('.txt','.hdf5')
+        if sparse:
+            outfilename = outfilename.replace('.hdf5','_sparse.hdf5')
+        f = h5py.File(outfilename, rdcc_nbytes=chunkSize, mode='w')
+
+        #save some lists of strings to the file for later use
+        hprocs = f.create_dataset("hprocs", [len(procs)], dtype=h5py.special_dtype(vlen=str), compression="gzip")
```

Comment:
Can't this be turned into a loop somehow...?

Reply:
Didn't make it a loop, but simpliefied